### PR TITLE
feat: add async inner to `BatchLogIteratorAdapter`

### PR DESCRIPTION
**Pull request by Rachelint**
…and add its adapting logic to `BatchLogIteratorAdapter`.

# Which issue does this PR close?

Closes #

# Rationale for this change
 The `BatchLogIteratorAdapter` just support `BlockingLogIterator` as its inner. It is not enough, because in other wal implementation, the iterator return from `read_batch` may be originally async, the  `BatchLogIteratorAdapter` should take  it in consideration.

# What changes are included in this PR?
+ Define a new trait  `AsyncLogIterator`.
+ Add its adapting logic to `BatchLogIteratorAdapter`.

# Are there any user-facing changes?
None.

# How does this change test
By exist ut.

### CODE REVIEW

File: wal/src/manager.rs [hunk 1]
```diff
@@ -286,44 +278,41 @@ pub trait WalManager: Send + Sync + fmt::Debug + 'static {
...
+    Sync(Box<dyn SyncLogIterator>, Arc<Runtime>),
+    Async(Box<dyn AsyncLogIterator>),
+}
+
+/// Adapter to convert a sync/async iterator to a batch async iterator.
 #[derive(Debug)]
 pub struct BatchLogIteratorAdapter {
-    blocking_iter: Option<Box<dyn BlockingLogIterator>>,
-    runtime: Arc<Runtime>,
+    inner_iterator: Option<InnerIterator>,
```
Comment (ShiKaiWi) [inline wal/src/manager.rs:290]: Just call it iter.

File: wal/src/manager.rs [hunk 2]
```diff
@@ -286,44 +278,41 @@ pub trait WalManager: Send + Sync + fmt::Debug + 'static {
...
     batch_size: usize,
 }
 
 impl BatchLogIteratorAdapter {
-    pub fn new(
-        blocking_iter: Box<dyn BlockingLogIterator>,
-        runtime: Arc<Runtime>,
-        batch_size: usize,
-    ) -> Self {
+    pub fn new(inner_iterator: InnerIterator, batch_size: usize) -> Self {
```
Comment (ShiKaiWi) [inline wal/src/manager.rs:295]: Maybe we should define two methods: `new_sync` & `new_async`.

File: wal/src/manager.rs [hunk 3]
```diff
@@ -286,44 +278,41 @@ pub trait WalManager: Send + Sync + fmt::Debug + 'static {
     async fn scan(&self, ctx: &ScanContext, req: &ScanRequest) -> Result<BatchLogIteratorAdapter>;
 }
 
-/// Adapter to convert a blocking interator to a batch async iterator.
+#[derive(Debug)]
+pub enum InnerIterator {
+    Sync(Box<dyn SyncLogIterator>, Arc<Runtime>),
```
Comment (ShiKaiWi) [inline wal/src/manager.rs:283]: Let's give the two fields names:
```rust
Sync {
    iter: Box<dyn SyncLogIterator>,
    runtime: Arc<Runtime>,
}
```

File: wal/src/manager.rs [hunk 4]
```diff
@@ -286,44 +278,41 @@ pub trait WalManager: Send + Sync + fmt::Debug + 'static {
     async fn scan(&self, ctx: &ScanContext, req: &ScanRequest) -> Result<BatchLogIteratorAdapter>;
 }
 
-/// Adapter to convert a blocking interator to a batch async iterator.
+#[derive(Debug)]
+pub enum InnerIterator {
```
Comment (ShiKaiWi) [inline wal/src/manager.rs:282]: How about naming it `BatchLogIterator`?

File: wal/src/manager.rs [hunk 5]
```diff
@@ -286,44 +278,41 @@ pub trait WalManager: Send + Sync + fmt::Debug + 'static {
...
             batch_size,
         }
     }
 }
 
-#[async_trait]
-impl BatchLogIterator for BatchLogIteratorAdapter {
-    async fn next_log_entries<D: PayloadDecoder + Send + 'static>(
+impl BatchLogIteratorAdapter {
+    async fn simulated_async_next<D: PayloadDecoder + Send + 'static>(
```
Comment (ShiKaiWi) [inline wal/src/manager.rs:304]: Add comments to explain what it means when `InnerIterator` is None.

File: wal/src/manager.rs (hunk 2 above)
Comment (Rachelint) [reply wal/src/manager.rs:295]: Yes, quite verbose to new it now.

File: wal/src/manager.rs (hunk 4 above)
Comment (Rachelint) [reply wal/src/manager.rs:282]: But the inner's return is not a batch, but is just a single log entry.
I propose to rename it to `LogIterator`.

File: wal/src/manager.rs (hunk 1 above)
Comment (Rachelint) [reply wal/src/manager.rs]: Done.

File: wal/src/manager.rs (hunk 2 above)
Comment (Rachelint) [reply wal/src/manager.rs]: Done.

File: wal/src/manager.rs (hunk 3 above)
Comment (Rachelint) [reply wal/src/manager.rs]: Done.

File: wal/src/manager.rs (hunk 4 above)
Comment (Rachelint) [reply wal/src/manager.rs]: Done.

File: wal/src/manager.rs (hunk 5 above)
Comment (Rachelint) [reply wal/src/manager.rs:319]: Done.

**Review by ShiKaiWi — APPROVED** [review]

**Pull request merged**